src/ssh.c

```diff
@@ -1479,137 +1476,231 @@ union wolfSSH_key {
-            word32 typeSz;
-            byte nameId;
-
-            typeSz = (word32)WSTRLEN(type);
-
-            nameId = NameToId(type, typeSz);
-            name = IdToName(nameId);
-            *outType = (const byte*)name;
-            *outTypeSz = typeSz;
-
-            if (*out == NULL) {
-                /* set size based on sanity check in wolfSSL base64 decode
-                 * function */
-                *outSz = ((word32)WSTRLEN(key) * 3 + 3) / 4;
-                newKey = (byte*)WMALLOC(*outSz, heap, DYNTYPE_PRIVKEY);
-                if (newKey == NULL) {
-                    return WS_MEMORY_E;
-                }
-                *out = newKey;
-            }
+    if (type != NULL && key != NULL) {
+        const char* name;
+        word32 typeSz;
+        byte nameId;
 
-            ret = Base64_Decode((byte*)key, (word32)WSTRLEN(key), *out, outSz);
-        }
+        typeSz = (word32)WSTRLEN(type);
 
-        if (ret != 0) {
-            WLOG(WS_LOG_DEBUG, "Base64 decode of public key failed.");
-            ret = WS_PARSE_E;
-        }
+        nameId = NameToId(type, typeSz);
+        name = IdToName(nameId);
+        *outType = (const byte*)name;
+        *outTypeSz = typeSz;
 
-        WFREE(c, heap, DYNTYPE_STRING);
-    }
-    else if (format == WOLFSSH_FORMAT_ASN1) {
         if (*out == NULL) {
-            newKey = (byte*)WMALLOC(inSz, heap, DYNTYPE_PRIVKEY);
+            /* set size based on sanity check in wolfSSL base64 decode
+             * function */
+            *outSz = ((word32)WSTRLEN(key) * 3 + 3) / 4;
+            newKey = (byte*)WMALLOC(*outSz, heap, DYNTYPE_PRIVKEY);
             if (newKey == NULL) {
                 return WS_MEMORY_E;
             }
             *out = newKey;
         }
-        else {
-            if (*outSz < inSz) {
-                WLOG(WS_LOG_DEBUG, "DER private key output size too small");
-                return WS_BUFFER_E;
-            }
-            newKey = *out;
+
+        ret = Base64_Decode((byte*)key, (word32)WSTRLEN(key), *out, outSz);
+    }
+
+    if (ret != 0) {
+        WLOG(WS_LOG_DEBUG, "Base64 decode of public key failed.");
+        ret = WS_PARSE_E;
+    }
+
+    WFREE(c, heap, DYNTYPE_STRING);
+    return ret;
+}
+
+
+static int DoAsn1Key(const byte* in, word32 inSz, byte** out,
+        word32* outSz, const byte** outType, word32* outTypeSz,
+        void* heap)
+{
+    int ret = WS_SUCCESS;
+    byte* newKey = NULL;
+
+    WOLFSSH_UNUSED(heap);
+
+    if (*out == NULL) {
+        newKey = (byte*)WMALLOC(inSz, heap, DYNTYPE_PRIVKEY);
+        if (newKey == NULL) {
+            return WS_MEMORY_E;
+        }
+        *out = newKey;
+    }
+    else {
+        if (*outSz < inSz) {
+            WLOG(WS_LOG_DEBUG, "DER private key output size too small");
+            return WS_BUFFER_E;
+        }
+        newKey = *out;
+    }
+    *outSz = inSz;
+    WMEMCPY(newKey, in, inSz);
+
+    ret = IdentifyAsn1Key(in, inSz, 1, heap);
+    if (ret > 0) {
```

Comment:
should we be free'ing newKey if ret is less than 0 here?

Reply:
Check.